### PR TITLE
Add snapshot-controller

### PR DESCRIPTION
**Pull request by olemarkus**
Will install external-snapshotter
Also modifies AWS storage to use a csi-specific storageclass and default to gp3 by default if ebs csi driver is installed.

### CODE REVIEW

File: upup/models/cloudup/resources/addons/storage-aws.addons.k8s.io/v1.15.0.yaml.template [hunk 1]
```diff
@@ -29,7 +29,7 @@ kind: StorageClass
 metadata:
   name: kops-ssd-1-17
   annotations:
-    storageclass.beta.kubernetes.io/is-default-class: "true"
+    storageclass.beta.kubernetes.io/is-default-class: "{{ if .CloudConfig.AWSEBSCSIDriver.Enabled }}{{ else }}false{{ end }}"
```
Comment (seh) [inline upup/models/cloudup/resources/addons/storage-aws.addons.k8s.io/v1.15.0.yaml.template:32]: I have a patch in the works that will require some coordination here, both with the annotation name and with surrounding the creation of the _StorageClass_ objects in a conditional block.

Comment (joshbranham) [reply upup/models/cloudup/resources/addons/storage-aws.addons.k8s.io/v1.15.0.yaml.template:32]: Yeah I really dislike the annotation path here for controlling which is the default. It made rolling out some best practices for new clusters + not breaking old oddly difficult back when I did this.

Comment (seh) [reply upup/models/cloudup/resources/addons/storage-aws.addons.k8s.io/v1.15.0.yaml.template:32]: See #10733 for my patch on the subject. We'll have to resolve a bunch of conflicts, depending on which one goes in first.

Comment (olemarkus) [reply upup/models/cloudup/resources/addons/storage-aws.addons.k8s.io/v1.15.0.yaml.template:32]: We'll merge yours first

File: upup/models/cloudup/resources/addons/storage-aws.addons.k8s.io/v1.15.0.yaml.template [hunk 2]
```diff
@@ -16,7 +15,7 @@ kind: StorageClass
 metadata:
   name: gp2
   annotations:
-    storageclass.kubernetes.io/is-default-class: "false"
+    storageclass.beta.kubernetes.io/is-default-class: "false"
```
Comment (rifelpet) [inline upup/models/cloudup/resources/addons/storage-aws.addons.k8s.io/v1.15.0.yaml.template:18]: is this addition still needed? I see upstream docs refer to the non-beta annotation.

**Comment by rifelpet** [comment]
this looks good to me other than the integration test conflicts. I'll let someone else take a second pass at the new API fields.

We should revert https://github.com/kubernetes/kops/pull/11530 in this PR as well.

File: pkg/model/components/awsebscsidriver.go [hunk 3]
```diff
@@ -31,10 +31,11 @@ var _ loader.OptionsBuilder = &AWSEBSCSIDriverOptionsBuilder{}
 
 func (b *AWSEBSCSIDriverOptionsBuilder) BuildOptions(o interface{}) error {
 	clusterSpec := o.(*kops.ClusterSpec)
-	c := clusterSpec.CloudConfig.AWSEBSCSIDriver
-	if c == nil {
-		return nil
+	cc := clusterSpec.CloudConfig
+	if cc.AWSEBSCSIDriver == nil {
```
[the remark this answers is not in the document]
Comment (olemarkus) [reply pkg/model/components/awsebscsidriver.go:35]: This would fail the tests if we had to, I think. This probably gets explicitly set in one of the former builders.

File: pkg/model/iam/iam_builder.go [hunk 4]
```diff
@@ -780,6 +783,19 @@ func AddAWSLoadbalancerControllerPermissions(p *Policy, resource stringorslice.S
...
+func addSnapshotPersmissions(p *Policy) {
+	p.Statement = append(p.Statement, &Statement{
+		Effect: StatementEffectAllow,
+		Action: stringorslice.Of(
+			"ec2:CreateSnapshot",
+			"ec2:DeleteSnapshot",
+			"ec2:DescribeAvailabilityZones",
+			"ec2:DescribeSnapshots",
+		),
+		Resource: stringorslice.Slice([]string{"*"}),
```
[the remark this answers is not in the document]
Comment (olemarkus) [reply pkg/model/iam/iam_builder.go:795]: Yep. that makes sense.

File: upup/models/cloudup/resources/addons/storage-aws.addons.k8s.io/v1.15.0.yaml.template (hunk 2 above)
Comment (olemarkus) [reply upup/models/cloudup/resources/addons/storage-aws.addons.k8s.io/v1.15.0.yaml.template]: This was a mistake. Reverted.

**Comment by olemarkus** [comment]
> this looks good to me other than the integration test conflicts. I'll let someone else take a second pass at the new API fields.
> 
> We should revert #11530 in this PR as well.

I want to cherry-pick this one to 1.21, so cannot revert that change directly in this PR. In addition the test need to be further modified to install the snapshot controller.

**Comment by rifelpet** [comment]
Looks like this still needs a rebase to get rid of a merge commit.

File: tests/integration/update_cluster/aws-lb-controller/data/aws_launch_template_master-us-test-1a.masters.minimal.example.com_user_data [hunk 5]
```diff
@@ -121,6 +121,8 @@ ensure-install-dir
 
 cat > conf/cluster_spec.yaml << '__EOF_CLUSTER_SPEC'
 cloudConfig:
+  awsEBSCSIDriver:
+    version: v1.0.0
```
Comment (rifelpet) [inline tests/integration/update_cluster/aws-lb-controller/data/aws_launch_template_master-us-test-1a.masters.minimal.example.com_user_data:125]: are we enabling the driver by default? I'm surprised to see the version set here if its not enabled.

Comment (olemarkus) [reply tests/integration/update_cluster/aws-lb-controller/data/aws_launch_template_master-us-test-1a.masters.minimal.example.com_user_data:125]: Spotted this one too. No, we are not. It should look better now.

File: upup/models/cloudup/resources/addons/storage-aws.addons.k8s.io/v1.15.0.yaml.template [hunk 6]
```diff
@@ -40,6 +40,20 @@ parameters:
 allowVolumeExpansion: true
 volumeBindingMode: WaitForFirstConsumer
 
+{{ if .CloudConfig.AWSEBSCSIDriver.Enabled }}
+---
+kind: StorageClass
+apiVersion: storage.k8s.iLoad Bao/v1
```
Comment (rifelpet) [inline upup/models/cloudup/resources/addons/storage-aws.addons.k8s.io/v1.15.0.yaml.template:46]:
```suggestion
apiVersion: storage.k8s.io/v1
```

File: tests/integration/update_cluster/aws-lb-controller/data/aws_launch_template_master-us-test-1a.masters.minimal.example.com_user_data [hunk 7]
```diff
@@ -121,6 +121,8 @@ ensure-install-dir
 
 cat > conf/cluster_spec.yaml << '__EOF_CLUSTER_SPEC'
 cloudConfig:
+  awsEBSCSIDriver:
+    enabled: false
```
Comment (rifelpet) [inline tests/integration/update_cluster/aws-lb-controller/data/aws_launch_template_master-us-test-1a.masters.minimal.example.com_user_data:125]: I'm wondering if we can exclude these fields altogether if the driver isn't enabled, and have the consumer of this data check for non nil. That way we aren't encroaching on userdata space

File: upup/models/cloudup/resources/addons/storage-aws.addons.k8s.io/v1.15.0.yaml.template [hunk 8]
```diff
@@ -40,6 +40,20 @@ parameters:
 allowVolumeExpansion: true
 volumeBindingMode: WaitForFirstConsumer
 
+{{ if WithDefaultBool .CloudConfig.AWSEBSCSIDriver.Enabled false }}
+---
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: kops-csi-1-20
```
Comment (rifelpet) [inline upup/models/cloudup/resources/addons/storage-aws.addons.k8s.io/v1.15.0.yaml.template:48]: sorry one more :/ 
```suggestion
  name: kops-csi-1-21
```

Comment (olemarkus) [reply upup/models/cloudup/resources/addons/storage-aws.addons.k8s.io/v1.15.0.yaml.template]: Thanks for catching :)

**Review by rifelpet — APPROVED** [review]
for a followup PR, is there anything we need to do with the existing default StorageClass?

https://github.com/kubernetes/kops/blob/01138a3995c28e98e29ba25d4314a882704c396a/upup/models/cloudup/resources/addons/storage-aws.addons.k8s.io/v1.15.0.yaml.template#L28-L41

Like some API validation around `CloudConfig.ManageStorageClasses` and `AWSEBSCSIDriver.Enabled` ? Or update that template to not set kops-ssd-1-17 as default if `AWSEBSCSIDriver.Enabled` ?

**Comment by rifelpet** [comment]
looks like nodeup has a panic:

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kops/10730/pull-kops-e2e-kubernetes-aws/1395461870592200704/artifacts/35.166.45.226/journal.log

```
May 20 19:37:26.067358 ip-172-20-33-37 nodeup[1325]: panic: runtime error: invalid memory address or nil pointer dereference
May 20 19:37:26.067415 ip-172-20-33-37 nodeup[1325]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x301ef1b]
May 20 19:37:26.067415 ip-172-20-33-37 nodeup[1325]: goroutine 1 [running]:
May 20 19:37:26.067494 ip-172-20-33-37 nodeup[1325]: k8s.io/kops/nodeup/pkg/model.(*AWSEBSCSIDriverBuilder).Build(0xc00000f588, 0xc000939ad0, 0x0, 0x0)
May 20 19:37:26.067494 ip-172-20-33-37 nodeup[1325]:         nodeup/pkg/model/awsebscsidriver.go:39 +0x5b
May 20 19:37:26.067574 ip-172-20-33-37 nodeup[1325]: k8s.io/kops/upup/pkg/fi/nodeup.(*Loader).Build(0xc000849d20, 0xc00001e600, 0x20, 0x20)
May 20 19:37:26.067574 ip-172-20-33-37 nodeup[1325]:         upup/pkg/fi/nodeup/loader.go:39 +0x344
May 20 19:37:26.067654 ip-172-20-33-37 nodeup[1325]: k8s.io/kops/upup/pkg/fi/nodeup.(*NodeUpCommand).Run(0xc00003c690, 0x4349cc0, 0xc00000e018, 0x0, 0x0)
May 20 19:37:26.067654 ip-172-20-33-37 nodeup[1325]:         upup/pkg/fi/nodeup/command.go:314 +0x2545
May 20 19:37:26.067654 ip-172-20-33-37 nodeup[1325]: main.main()
May 20 19:37:26.067654 ip-172-20-33-37 nodeup[1325]:         cmd/nodeup/main.go:117 +0x1012
```

**Comment by olemarkus** [comment]
Classic mistake of only checking if the struct exists, not if it was actually enabled.

**Review by hakman — APPROVED** [review]

**Comment by rifelpet** [comment]
/hold
the default storage class logic in the template needs to be inverted

**Comment by olemarkus** [comment]
Also added encryption back in

**Review by rifelpet — APPROVED** [review]
/hold cancel

**Review by rifelpet — APPROVED** [review]

**Review by rifelpet — APPROVED** [review]
